src/deadline/job_attachments/_diff.py

```diff
@@ -0,0 +1,110 @@
+from deadline.client.exceptions import NonValidInputError
+from deadline.job_attachments.asset_manifests.base_manifest import BaseAssetManifest
+from deadline.job_attachments.caches.hash_cache import HashCache
+from deadline.job_attachments.models import AssetRootManifest, FileStatus
+from deadline.job_attachments.upload import S3AssetManager
+
+
+def diff_manifest(
+    asset_manager: S3AssetManager,
+    asset_root_manifest: AssetRootManifest,
+    manifest: str,
+    update: bool,
+) -> List[tuple]:
+    """
+    Gets the file paths in specified manifest if the contents of file have changed since its last snapshot.
+    """
+    manifest_dir_name = os.path.basename(manifest)
+    root_path = asset_root_manifest.root_path
+    input_paths: List[Path] = []
+
+    asset_manifest = asset_root_manifest.asset_manifest
+    if asset_manifest is None:
+        raise NonValidInputError("Manifest object not found, please check input manifest. ")
+
+    for base_manifest_path in asset_manifest.paths:
+        if base_manifest_path.path.startswith(manifest_dir_name):
+            # skip the manifest folder, or else every upload will need an update after a previous change
+            continue
+        input_paths.append(Path(root_path, base_manifest_path.path))
+
+    return find_file_with_status(
+        asset_manager=asset_manager,
+        input_paths=input_paths,
+        root_path=root_path,
+        update=update,
+        statuses=[FileStatus.NEW, FileStatus.MODIFIED],
+    )
+
+
+def find_file_with_status(
+    asset_manager: S3AssetManager,
+    input_paths: List[Path],
+    root_path: str,
+    update: bool,
+    statuses: List[FileStatus],
+) -> List[tuple]:
+    """
+    Checks a manifest file, compares it to specified root directory or manifest of files with the local hash cache, and finds files that match the specified statuses.
+    Returns a list of tuples containing the file information, and its corresponding file status.
+    """
+    cache_config = config_file.get_cache_directory()
+
+    with HashCache(cache_config) as hash_cache:
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            futures = {
+                executor.submit(
+                    asset_manager._process_input_path,
+                    path=path,
+                    root_path=root_path,
+                    hash_cache=hash_cache,
+                    update=update,
+                ): path
+                for path in input_paths
+            }
+            status_paths: List[tuple] = []
+            for future in concurrent.futures.as_completed(futures):
+                (file_status, _, manifestPath) = future.result()
+                if file_status in statuses:
+                    status_paths.append((file_status, manifestPath))
+
+            return status_paths
+
+
+def compare_manifest(
+    reference_manifest: BaseAssetManifest, compare_manifest: BaseAssetManifest
+) -> List[(tuple)]:
```

Comment:
nit - maybe `List[Tuple[FileStatus, BaseManifest]]`?

Reply:
Fixed! I think this one I removed the static typing due to `all:lint`, but have added it back.